Sources/Testing/Attachments/Test.Attachable.swift

```diff
@@ -0,0 +1,72 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+//
+
+@_spi(Experimental)
+extension Test {
+  /// A protocol describing a type that can be attached to a test report or
+  /// written to disk when a test is run.
+  ///
+  /// To attach an attachable value to a test report or test run output, use it
+  /// to initialize a new instance of ``Test/Attachment``, then call
+  /// ``Test/Attachment/attach()``. An attachment can only be attached once.
+  ///
+  /// Generally speaking, you should not need to make new types conform to this
+  /// protocol.
+  // TODO: write more about this protocol, how it works, and list conforming
+  // types (including discussion of the Foundation cross-import overlay.)
+  public protocol Attachable: ~Copyable {
+    /// Call a function and pass a buffer representing this instance to it.
+    ///
+    /// - Parameters:
+    ///   - attachment: The attachment that is requesting a buffer (that is, the
+    ///     attachment containing this instance.)
+    ///   - body: A function to call. A temporary buffer containing a data
+    ///     representation of this instance is passed to it.
+    ///
+    /// - Returns: Whatever is returned by `body`.
+    ///
+    /// - Throws: Whatever is thrown by `body`, or any error that prevented the
+    ///   creation of the buffer.
+    ///
+    /// The testing library uses this function when writing an attachment to a
+    /// test report or to a file on disk. The format of the buffer is
+    /// implementation-defined, but should be "idiomatic" for this type: for
+    /// example, if this type represents an image, it would be appropriate for
+    /// the buffer to contain an image in PNG format, JPEG format, etc., but it
+    /// would not be idiomatic for the buffer to contain a textual description
+    /// of the image.
+    borrowing func withUnsafeBufferPointer<R>(for attachment: borrowing Test.Attachment, _ body: (UnsafeRawBufferPointer) throws -> R) throws -> R
+  }
+}
+
+// MARK: - Default implementations
+
+// Implement the protocol requirements for byte arrays and buffers so that
+// developers can attach raw data when needed.
+@_spi(Experimental)
+extension [UInt8]: Test.Attachable {
```

Comment:
Can we also make String conform to Test.Attachable?

Reply:
Short answer "yes, trivially so", slightly longer answer "are we sure we want to?" Strings are a currency type and don't provide any sort of metadata, so we'd want to gently guide developers away from them if there's a better data type to use.

But yes, I should add conformance for `String` and `Substring`.

Reply:
I'm not a huge fan of the Attachable conformance requiring developers to implement `unsafe*` methods because it's a) complicated, b) can be, well, unsafe and c) looks bad. If we can conform common types to it, that would shield developers from having to handle the footgun. If we can refine the interface even better.

Reply:
I suspect once we get `Span` in the language, we'll want to use that instead of `unsafe*`, but since the currency type in Swift for "bag of bytes" today is `UnsafeRawBufferPointer` (or `UnsafeBufferPointer<UInt8>`), that's what we need the interface to use.

Reply:
Also in a subsequent PR I add default conformances for `Encodable` and `NSSecureCoding`, so most types can get conformance "for free" if they're already codable.

Reply:
Great, that's what I'm after, a default conformance on Encodable/NSSecureCoding will go a long way.